### ISSUE TITLE
add support to directly execute statements - EXECUTE IMMEDIATE

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#50]
https://docs.oracle.com/cd/A97630_01/appdev.920/a96109/pco09dyn.htm#774
https://www.postgresql.org/docs/14/ecpg-dynamic.html

```cobol
EXEC SQL
    EXECUTE IMMEDIATE 'some hard-coded sql here'
END-EXEC
EXEC SQL
    EXECUTE IMMEDIATE :stmt
END-EXEC
```

As a wide-supported way to execute something that has no bind parameters (only the statement if not given by literal) and no result set (The `SQLCA` structure would still be set, so checking`SQLCODE` and friends still would be possible).

**Comment by GitMensch** [comment]
Hm, checking the code makes me wonder: it doesn't look like `EXEC SQL PREPARE prepared_stname FROM :statement_id` (or literal) and following `EXEC SQL EXECUTE prepared_stname` isn't supported either, is it?
If not do we want to track it here or add a separate issue?

[the execution of previously binded statements to be able to toggle db options is actually one of the more important things that _may_ be missing for me, the additional "IMMEDIATE" way would be a bonus only)]

**Comment by GitMensch** [comment]
... Rechecked with "upstream" ocesql: they likely added it after your fork but have an issue that the parser can _only_ handle variable-length items in the statement as variable length group (ideally both would be supported, taking the length either explicit from the definition or - in the group case - from the `xyz-LEN` variable).

**Comment by mridoni** [comment]
I'll have a look at the documentation, and see how this can be implemented. The EXECUTE IMMEDIATE is quite straightforward (it's basically a passthrough from COBOL to GixSQLExec) but prepared statements are obviously more complicated.

P.S. about your comment above("...the parser can only handle variable-length items in the statement as variable length group..."): are you sure this is related to this issue and not to a different one? I cannot understand the relation.

**Comment by GitMensch** [comment]
>about your comment above("...the parser can only handle variable-length items in the statement as variable length group..."): are you sure this is related to this issue and not to a different one? I cannot understand the relation.

That was just a side note about ocesql's implementation of it (a bug there)

**Comment by mridoni** [comment]
Ok, I have started some work on the "prepare statement" feature; obviously at this stage I cannot give a timeline, but I should know something more as I progress, probably by the end of the week. For now I intend to cover "method" 1 and 2 in the Oracle docs you linked, leaving cursor support aside for the time being.

**Comment by GitMensch** [comment]
That sounds like a reasonable start (but wouldn't the cursor stuff be identical to the normal cursor fetching)?

Note: cursor fetching is also implemented in ocesql and if you leave the "normal" errors (those that happen with any cursor featching) aside it works. #16 seems to come back, I highly suggest to look at it to possibly get the "base" of dynamic sql covered identical and working from there instead of completely re-implement it.

And a side note from the linked postgresql docs: I've never seen it before but a `EXEC SQL DEALLOCATE PREPARE name` may be reasonable (ocesql does not have that, but it checks "was the name already used - then issue a warning [which is only seen on debug logging], free everything related to that, then allocate from the beginning".

**Comment by mridoni** [comment]
> but wouldn't the cursor stuff be identical to the normal cursor fetching?

Yes, once I have the prepared statement "stuff" in place, it shouldn't be difficult to add support for cursors, but I would like to proceed gradually. Moreover, I also need to deal with supporting prepared statements in ODBC and MySQL, and there are a couple of things that are are currently only supported in PostgreSQL, so I need to realign the drivers in terms of features, though maybe (for now) not completely.

> I highly suggest to look at it to possibly get the "base" of dynamic sql covered identical and working from there instead of completely re-implement it.

Thanks for the suggestion, I'll surely look into it.

**Comment by mridoni** [comment]
I looked at how ocesql implements prepared statements at runtime: nothing fancy, just some string splitting and substitution, it's more or less a copy of the algorithm used by the parser to deal with parameterized statements. The code is actually quite smart, but it needs to be, given that it is pure C, without string manipulation facilities. Anyway, I will implement the same behavior in GixSQL.

**Comment by mridoni** [comment]
It is coming along nicely. This is the test case I am using:

       IDENTIFICATION DIVISION.
       
       PROGRAM-ID. TSQL006A. 
       
       
       ENVIRONMENT DIVISION. 
       
       CONFIGURATION SECTION. 
       SOURCE-COMPUTER. IBM-AT. 
       OBJECT-COMPUTER. IBM-AT. 
       
       INPUT-OUTPUT SECTION. 
       FILE-CONTROL. 
       
       DATA DIVISION.  
       
       FILE SECTION.  
       
       WORKING-STORAGE SECTION. 
       
           01 DATASRC PIC X(64).
           01 DBUSR  PIC X(64).

           01  DYNSTMT1   SQL TYPE IS VARCHAR(100).
           01  DNUM1   PIC S9(4)V99.
           01  DNUM2   PIC S9(4)V99.
           01  DNUM3   PIC S9(4)V99.
           01  T1      PIC 9(4) VALUE 3.
           01  T2      PIC 9(4) VALUE 4.
           01  TOT1    PIC 9(4) VALUE 0.
           01  TOT2    PIC 9(4) VALUE 0.

       EXEC SQL 
            INCLUDE SQLCA 
       END-EXEC. 
         
       PROCEDURE DIVISION. 
 
       000-CONNECT.

           DISPLAY "DATASRC" UPON ENVIRONMENT-NAME.
           ACCEPT DATASRC FROM ENVIRONMENT-VALUE.
           DISPLAY "DATASRC_USR" UPON ENVIRONMENT-NAME.
           ACCEPT DBUSR FROM ENVIRONMENT-VALUE.
           
           DISPLAY '***************************************'.
           DISPLAY " DATASRC  : " DATASRC.
           DISPLAY " AUTH     : " DBUSR.
           DISPLAY '***************************************'.

           EXEC SQL
              CONNECT TO :DATASRC USER :DBUSR
           END-EXEC.      
           
           IF SQLCODE <> 0 THEN
              DISPLAY 'CONNECT SQLCODE. ' SQLCODE
              DISPLAY 'CONNECT SQLERRM. ' SQLERRM
              GO TO 100-EXIT
           END-IF.

       100-MAIN.

           EXEC SQL START TRANSACTION END-EXEC.


           EXEC SQL DROP TABLE IF EXISTS TAB1 END-EXEC.
           DISPLAY 'CONNECT DROP(1): ' SQLCODE

           EXEC SQL CREATE TABLE TAB1 (FLD1 INT, FLD2 INT) END-EXEC.
           DISPLAY 'CONNECT CREATE(1): ' SQLCODE

          MOVE 'INSERT INTO TAB1(FLD1, FLD2) VALUES(?, ?)' TO DYNSTMT1.
           
           EXEC SQL 
               PREPARE SQLSTMT1 FROM :DYNSTMT1
           END-EXEC.
           DISPLAY 'PREPARE SQLSTMT1: ' SQLCODE

           MOVE 1 TO T1.
           MOVE 2 TO T2.
           EXEC SQL EXECUTE SQLSTMT1 USING :T1, :T2 END-EXEC.
           DISPLAY 'EXECUTE SQLSTMT1(1): ' SQLCODE

           MOVE 5 TO T1.
           MOVE 6 TO T2.
           EXEC SQL EXECUTE SQLSTMT1 USING :T1, :T2 END-EXEC.
           DISPLAY 'EXECUTE SQLSTMT1(2): ' SQLCODE

           EXEC SQL EXECUTE IMMEDIATE 
                'UPDATE TAB1 SET FLD1=FLD1+100, FLD2=FLD2+300'
           END-EXEC.
           DISPLAY 'EXECUTE IMMEDIATE(1): ' SQLCODE

           MOVE 'UPDATE TAB1 SET FLD1=FLD1+100, FLD2=FLD2+300'
      -          TO DYNSTMT1.

           EXEC SQL EXECUTE IMMEDIATE :DYNSTMT1 END-EXEC.
           DISPLAY 'EXECUTE IMMEDIATE(2): ' SQLCODE

           EXEC SQL
                SELECT SUM(FLD1), SUM(FLD2)
                    INTO :TOT1, :TOT2 FROM TAB1
           END-EXEC
           DISPLAY 'SUM SQLCODE: ' SQLCODE.
           
           DISPLAY 'TOT1: ' TOT1.
           DISPLAY 'TOT2: ' TOT2.

           EXEC SQL COMMIT END-EXEC.

           EXEC SQL
              CONNECT RESET
           END-EXEC.      
           
           IF SQLCODE <> 0 THEN
              DISPLAY 'DISCONNECT SQLCODE. ' SQLCODE
              DISPLAY 'DISCONNECT SQLERRM. ' SQLERRM
              GO TO 100-EXIT
           END-IF.
       
       100-EXIT. 
             STOP RUN.

 that passes with the following (correct) output:

    ***************************************
     DATASRC  : pgsql://localhost:5432/testdb1                                  
     AUTH     : test.test                                                       
    ***************************************
    CONNECT DROP(1): +0000000000
    CONNECT CREATE(1): +0000000000
    PREPARE SQLSTMT1: +0000000000
    EXECUTE SQLSTMT1(1): +0000000000
    EXECUTE SQLSTMT1(2): +0000000000
    EXECUTE IMMEDIATE(1): +0000000000
    EXECUTE IMMEDIATE(2): +0000000000
    SUM SQLCODE: +0000000000
    TOT1: 0406
    TOT2: 1208

I need to triage the remaining issues I am going to incorporate in this release, then start planning for release, Version 1.0.10 should be out, barring any unforeseen issues or, as usual constraints due to my day job, by the end of the next week.

**Comment by GitMensch** [comment]
Looks good to me. The only part I've wondered was the missing `SQL DECLARE SECTION` - What effect does it have with gixsql? Note: if it isn't needed then the side-effect is that gixpp would have to be able to parse _every_ preprocessor code or "must come last" in the call chain of preprocessors.

Please have a look at #58 in any case (that's even related to the question above :-) and the pull-requests when doing triage. Thank you.

**Comment by mridoni** [comment]
Currently SQL DECLARE SECTION is basically a no-op in GixSQL. Or it should be, I will look at #58.

Thanks

**Comment by GitMensch** [comment]
> Currently SQL DECLARE SECTION is basically a no-op in GixSQL.

Doesn't this mean that the preprocessor must be able to "parse everything", including exotic `SCREEN SECTION` extensions and `REPORT WRITER SECTION`?

**Comment by mridoni** [comment]
It should jump over them, but I will test this further and see how it currently behaves, then one can move from there.

**Comment by mridoni** [comment]
Nope, need to fix this, thanks for pointing it out

**Comment by GitMensch** [comment]
> It should jump over them

Over what? `SCREEN SECTION`, `REPORT WRITER SECTION` and any unknown `USAGE`?
The last may happen in any case but especially likely to happen with `TYPEDEF`s defined in a copybook.

Either the preprocessor must know how to handle (or skip) all of those (then likely also things like `OCCURS UNBOUNDED`) and conditional compilation via `>> IF`/`$IF` or skip everything else out of `PROCEDURE DIVISION` than `EXEC SQL DECLARE`.

**Comment by mridoni** [comment]
Point taken, I am looking into it.

**Comment by GitMensch** [comment]
I agree that in the medium term it would be good to split the parser (all `EXEC` going somewhere specific), and also agree that this is not something that would have to be done in the next weeks/months. As this was about `EXECUTE IMMEDIATE` I suggest this issue be closed after the next release, I'm creating a new one for the parser split.

Concerning that side issue: have you checked that the parser does not stumble over `USAGE stuff-it-does-not-recognize`? We heavily use TYPEDEFs (outside of `DECLARE SECTION`).

**Comment by mridoni** [comment]
>> Concerning that side issue: have you checked that the parser does not stumble over USAGE stuff-it-does-not-recognize? We heavily use TYPEDEFs (outside of DECLARE SECTION).

Can you provide me with some snippets (or docs) I can use for testing?

**Comment by GitMensch** [comment]
> Can you provide me with some snippets (or docs) I can use for testing?

Sure, here's a sample that was provided in the GnuCOBOL FR to implementing this:
https://sourceforge.net/p/gnucobol/feature-requests/193/#5e11

and here are the test cases from GnuCOBOL's testsuite (first: Micro Focus format, then COBOL2002 format)

```cobol
AT_DATA([prog.cob], [
       IDENTIFICATION   DIVISION.
       PROGRAM-ID.      prog.
       DATA             DIVISION.
       WORKING-STORAGE  SECTION.
       01 AUSGABE-FILE-NAME-T          PIC X(50) IS TYPEDEF.
       01 SOME-VERY-LONG-TYPEDEF-NAME  PIC 9999  IS TYPEDEF.
       01 AUSGABE-FILE-NAME-2T         IS TYPEDEF.
          05 FILLER    PIC 9999.
          05 DETAIL-NO USAGE SOME-VERY-LONG-TYPEDEF-NAME.
      *
       01  MESSAGE-TEXT-2T IS TYPEDEF.
         02 AUSGABE-FILE-NAME     USAGE AUSGABE-FILE-NAME-T.
         02 FILLER REDEFINES AUSGABE-FILE-NAME.
             05 FILLER PIC 9999.
         02 AUSGABE-FILE-NAME-2   USAGE AUSGABE-FILE-NAME-2T.
         02 FILLER                USAGE AUSGABE-FILE-NAME-T.
      *
       01  MESSAGE-TEXT-2 EXTERNAL USAGE MESSAGE-TEXT-2T.
      
       77 OUTPUT-NAME USAGE SOME-VERY-LONG-TYPEDEF-NAME GLOBAL.

       01 Z-MESSAGE-T2 USAGE AUSGABE-FILE-NAME-2T.
       01 Z-MESSAGE-T3.
          49 MT3                    USAGE MESSAGE-TEXT-2T.
          49 MT3-REN  REDEFINES MT3 USAGE MESSAGE-TEXT-2T.

       77 CALCULUS             PIC S9(15)V9(03) IS TYPEDEF.
       01 USER-TYPE            IS TYPEDEF.
          02 AMOUNT            USAGE CALCULUS.
          02 FILLER            OCCURS 100.
             05 GRP-AMOUNT     USAGE CALCULUS.
       01 USER-VAR             USAGE USER-TYPE.

       PROCEDURE DIVISION.
           DISPLAY AUSGABE-FILE-NAME OF MESSAGE-TEXT-2
           DISPLAY DETAIL-NO         OF Z-MESSAGE-T2
           DISPLAY AUSGABE-FILE-NAME OF MT3
           DISPLAY OUTPUT-NAME
           MOVE -123.45 TO AMOUNT
           MOVE AMOUNT  TO GRP-AMOUNT (1)
           GOBACK.
])

AT_DATA([progstd.cob], [
       IDENTIFICATION   DIVISION.
       PROGRAM-ID.      prog.
       DATA             DIVISION.
       WORKING-STORAGE  SECTION.
       01 AUSGABE-FILE-NAME-T     PIC X(50) IS TYPEDEF.
       01 DETAIL-NO-T             PIC 9999  IS TYPEDEF.
       01 AUSGABE-FILE-NAME-2T              IS TYPEDEF.
          05 FILLER    PIC 9999.
          05 DETAIL-NO TYPE TO DETAIL-NO-T.
      *
       01  MESSAGE-TEXT-2T IS TYPEDEF.
         02 AUSGABE-FILE-NAME     TYPE AUSGABE-FILE-NAME-T.
         02 FILLER REDEFINES AUSGABE-FILE-NAME.
             05 FILLER PIC 9999.
         02 AUSGABE-FILE-NAME-2   TYPE AUSGABE-FILE-NAME-2T.
         02 FILLER                TYPE AUSGABE-FILE-NAME-T.
      *
       01  MESSAGE-TEXT-2 EXTERNAL TYPE MESSAGE-TEXT-2T.
      
       77 OUTPUT-NAME TYPE TO DETAIL-NO-T GLOBAL.

       01 Z-MESSAGE-T2 TYPE AUSGABE-FILE-NAME-2T.
       01 Z-MESSAGE-T3.
          49 MT3                    TYPE MESSAGE-TEXT-2T.
          49 MT3-REN  REDEFINES MT3 TYPE MESSAGE-TEXT-2T.

       77 CALCULUS             PIC S9(15)V9(03) IS TYPEDEF.

       01 SOME-STRUCT          IS TYPEDEF.
          02 SOME-DATA         PIC 9.
          02 SOME-VALUES       OCCURS 3.
             03 OTHER-DATA     PIC X.
             03 OTHER-VALUES   OCCURS 2 PIC 9.
       77 SOME2-DATA           PIC 9.      *> temporary variant until solved...
       01 SOME-STRUCT2         IS TYPEDEF.
      *>  02 SOME2-DATA        PIC 9.      TODO!
          02 SOME2-VALUES      OCCURS 1 TO 6 DEPENDING ON SOME2-DATA
                               DESCENDING KEY SB SA
                               INDEXED BY SOME2-INDEX.
             05 SA             PIC X VALUE x'12'.
             05 SB             PIC 9 VALUE 0.
             05 SDATA          PIC X(12).

       01 MY-TEST.
          02 AMOUNT            TYPE CALCULUS.
          02 FILLER            OCCURS 100.
             05 GRP-AMOUNT     TYPE CALCULUS.
          02 MY-NAME           TYPE SOME-STRUCT  OCCURS 5.
          02 MY-NAME2          TYPE SOME-STRUCT2.
       01 MY-TEST2             TYPE SOME-STRUCT2.
       LINKAGE SECTION.
       01 MY-STORE             TYPE CALCULUS.

       PROCEDURE DIVISION USING MY-STORE.
           DISPLAY AUSGABE-FILE-NAME OF MESSAGE-TEXT-2
           DISPLAY DETAIL-NO         OF Z-MESSAGE-T2
           DISPLAY AUSGABE-FILE-NAME OF MT3
           DISPLAY OUTPUT-NAME
           SET ADDRESS OF MY-STORE TO NULL
           INITIALIZE MY-NAME(2) GRP-AMOUNT (99)
           INITIALIZE OTHER-VALUES (2, 3, 2)
           SEARCH ALL SOME2-VALUES OF MY-NAME2
      *>     WHEN SA(SOME2-INDEX) = x'12'  CHECKME: Should work
             WHEN SA OF MY-NAME2(SOME2-INDEX OF MY-NAME2) = x'12'
                CONTINUE.
           GOBACK.
])
```

As far as I know most preprocessor's don't know how to handle that, which is not a problem if they can skip those definitions (either completely because they are not part of `DECLARE SECTION` [for Open-COBOL ESQL it is necessary to additionally move the _definitions_ out to copybooks and use the MF format because it stumbles over `TYPE`] or by just setting to "unknown type") and don't need it because the variables involved are not used in the preparser.

GnuCOBOL claims full support for the MF variant (actually works like a charm) and initial support for the COBOL 2002 one (I don't know of production uses of that and related feedback).

**Comment by mridoni** [comment]
Ok, thanks, I'll look into this

**Comment by mridoni** [comment]
Implemented and passed initial tested, will be in v1.0.10

**Comment by GitMensch** [comment]
> Implemented and passed initial tested, will be in v1.0.10

What was implemented now? The original issue (as from Tuesday) and/or TYPEDEFs in general (this may should be split out, not sure) and/or "skipping any unknown USAGE without error (as long as it isn't used as a bind variable)"?

**Comment by mridoni** [comment]
`EXECUTE IMMEDIATE`, `EXECUTE PREPARED` (with parameters) and improved parsing/skipping of `SQL DECLARE`, `SCREEN` and `REPORT` sections will be in v1.0.10, sorry for the confusion. All the more advanced stuff (improved parser, handling typedefs, preprocessing directives and such) will have to wait until #63, that still needs some planning. Sorry for the confusion.

**Comment by GitMensch** [comment]
I'm stumbling over some code that gixpp cannot handle (especially within conditional compilation).
Would it be possible/reasonable to add something like `EXEC SQL COMMENT` 100  lines of code `EXEC SQL END` to be able to mark a code block as "don't parse anything in here" (= the first one starts a lexer case where everything is just copied "as is")?

**Comment by mridoni** [comment]
> Would it be possible/reasonable to add something like EXEC SQL COMMENT 100 lines of code EXEC SQL END to be able to mark a code block as "don't parse anything in here" (= the first one starts a lexer case where everything is just copied "as is")?

Yes, for now it would be a good solution. Would `EXEC SQL IGNORE` instead of `EXEC SQL COMMENT` be acceptable?

**Comment by GitMensch** [comment]
`IGNORE` was my first thought, too. Yes, I'm fine with that.

**Comment by GitMensch** [comment]
Recheched the prepared statement - works as long as the variable binded for the statement is plain PIC X.
With ocesql this did not work so it was adjusted to the _only_ thing working **there**: a VARYING / VARLENGTH group - and this does not work in GixSQL yet...

```cobol
       01  S-SQLCOMMAND.
           03 S-SQLCOMMAND-LEN    PIC S9(8) COMP-5.
           03 S-SQLCOMMAND-ARR    PIC X(250).

           MOVE EXEC-SQLCOMMAND TO S-SQLCOMMAND-ARR.
           MOVE FUNCTION LENGTH(FUNCTION TRIM(S-SQLCOMMAND-ARR))
             TO S-SQLCOMMAND-LEN.
             

           EXEC SQL AT :DBS
               PREPARE P1 FROM :S-SQLCOMMAND 
           END-EXEC.
```

translated to

```
OCESQL     CALL "OCESQLPrepare" USING
OCESQL          BY REFERENCE SQLCA
OCESQL          BY REFERENCE "P1" & x"00"
OCESQL          BY REFERENCE S-SQLCOMMAND-ARR
OCESQL          BY VALUE S-SQLCOMMAND-LEN
OCESQL     END-CALL.
```
vs.
```
GIXSQL     CALL "GIXSQLPrepareStatement" USING
GIXSQL         BY REFERENCE SQLCA
GIXSQL         BY REFERENCE DBS
GIXSQL         BY VALUE 10
GIXSQL         BY REFERENCE "P1" & x"00"
GIXSQL         BY REFERENCE S-SQLCOMMAND
GIXSQL         BY VALUE 0
GIXSQL     END-CALL
```

GIXSQLPrepareStatement does not check any type, it just uses the string at this place which is pointing to a COMP-3 variable, not to the actual statement...


There should be a compile-time check of the time and either an error, or preferably a type hint passed to and used in GIXSQLPrepareStatement.
Changing it to a plain ´PIC X` - and it works.

**Comment by GitMensch** [comment]
For both `EXECUTE IMMEDIATE` and prepared statements it seems necessary to also support (or explicit forbid, but in any case compile-time-check) the type of the binded parameter because of VARLEN parameters.
`EXECUTE IMMEDIATE` works like a charm with both literals and plain PIC X - did not tested VARLEN parameters there.

I think we should split this issue as it contains too much "related but different" entries.
Here's one part of the cursor declarations where the codegen is broken:

```sql
           EXEC SQL AT :DBS
              DECLARE CSKEY01N CURSOR FOR
                 SELECT COL1, COL2 FROM TAB
                 WHERE KEY01 >= :TABKEY
                 ORDER BY KEY01 ASC
           END-EXEC.
```

This generates a call to `GIXSQLCursorDeclareParams` but misses the last parameter.

I've tracked this down to TPESQLProcessing.cpp, _thinking_ it would be around line 536 - which is totally fine, but found the generation be done "dynamic" around line 807, where the host_list->size is not passed).

**Comment by mridoni** [comment]
> I think we should split this issue as it contains too much "related but different" entries.

Yes, I agree, it is becoming difficult to keep track of all the sub-issues.

**Comment by mridoni** [comment]
This (and #76) will go into 1.0.12, I wanted to publish first v1.0.11 to get rid of a few simpler things and proceed with these two issues that need more testing. In the next few weeks I will be busier with my day-job, so I cannot give a timeline.

**Comment by mridoni** [comment]
> I've tracked this down to TPESQLProcessing.cpp, _thinking_ it would be around line 536 - which is totally fine, but found the generation be done "dynamic" around line 807, where the host_list->size is not passed).

Ok, found. The problem is (was) just that. I didn't see it before because, without giving much thought to it, I was declaring all the cursors for my test cases in the ` WORKING-STORAGE` section. The code generation for cursors declared there and not in the `PROCEDURE DIVISION` follows a different (and correct) path, so I was always checking the working case and not the other. I fixed it like this: not using the call id as in your suggested fix, but the size of the vector where the parameters are stored, it is basically the same thing, but this way it is more consistent with the checks done above.

    Index: TPESQLProcessing.cpp
    ===================================================================
    --- TPESQLProcessing.cpp	(revision 1637)
    +++ TPESQLProcessing.cpp	(working copy)
    @@ -843,6 +843,8 @@
     				select_call.addParameter("\"" + stmt->cursorName + "\" & x\"00\"", BY_REFERENCE);
     				select_call.addParameter(stmt->cursor_hold, BY_VALUE);
     				select_call.addParameter(string_format("SQ%04d", stmt->sql_query_list_id), BY_REFERENCE);
    +				if (stmt->host_list->size())
    +					select_call.addParameter(stmt->host_list->size(), BY_VALUE);
     				
     				if (!put_call(select_call, stmt->period))
     					return false;

**Comment by mridoni** [comment]
Spoke too soon, found another bug that has to do with associating the cursor to the right connection, I will have to investigate further.

**Comment by mridoni** [comment]
> Recheched the prepared statement - works as long as the variable binded for the statement is plain PIC X.
With ocesql this did not work so it was adjusted to the only thing working there: a VARYING / VARLENGTH group - and this does not work in GixSQL yet...

Added type checking in the internal repository, this case (see the comment) is now trapped and an error raised. By the way: if you change the level of the two sub-fields from `03` to `49` it also works (the same as it did before).

**Comment by GitMensch** [comment]
Just found that level thing (again) when inspecting the code and will adjust that.

**Comment by GitMensch** [comment]
Looks like this can be closed, as "changed in the internal repo" and released.
